### DISCUSSION TITLE
Switches and Rotary Encoder Using a Resistor Ladder

### DISCUSSION THREAD

**Discussion by GoremanX** [gpiozero/gpiozero#1119] [Q&A]
I recently purchased a fancy schmancy control for a vehicle media player. It has dual concentric shafts:
[https://www.crutchfield.com/S-4MnIrGMnjiL/p_068258A/RetroSound-258A.html](url)

- inner shaft is a rotary encoder and push-button momentary switch
- outer shaft is a 3-position momentary rotary switch (onL-off-onR) that rests in the off position

These are used on RetroSound head units, but are also available separately from RetroSound as a service part. They have 8 terminals:
- 3 for the rotary encoder
- 2 for the momentary push button on the inner knob
- 3 for the momentary rotary switch action of the outer shaft

They're soldered to a small circuit board that has a resistor ladder. The wire that actually connects to this board only has 4 conductors instead of 8 (it's an RJ11 jack). So they managed to get the functions of 3 switches and 1 rotary encoder into just 4 connections.

However none of this is documented anywhere, and RetroSound isn't inclined to educate me, so I'm trying to figure out how this wiring works.

There's 4 resistors on the circuit board (see photo, attached):
- 47k ohm (smd code 473)
- 82k ohm (smd code 823)
- 6.8k ohm (smd code 682)
- 22k ohm (smd code 223)

![IMG-20240203-213637054](https://github.com/gpiozero/gpiozero/assets/876310/904b6d0d-d998-4c47-8c9a-5ade02c53ad0)

So far, I've figured out:

- touch nothing: open between <code style="color:yellow">**yellow**</code> and <code style="color : green">**green**</code> / <code style="color : white">**black**</code> and <code style="color : green">**green**</code>
- turn outer knob left: 6.8k ohms between <code style="color : yellow">**yellow**</code> and <code style="color : green">**green**</code> / <code style="color : white">**black**</code> and <code style="color : green">**green**</code>
- turn outer knob right: 22k ohms between <code style="color : yellow">**yellow**</code> and <code style="color : green">**green**</code> / <code style="color : white">**black**</code> and <code style="color : green">**green**</code>
- push inner knob: 0 ohms between <code style="color : yellow">**yellow**</code> and <code style="color : green">**green**</code> / <code style="color : white">**black**</code> and <code style="color : green">**green**</code>
- encoder turning right: starts open at detent, 82k ohms between detents, 30k ohms at detent, 47k ohms between detents, back to open at detent between <code style="color : red">**red**</code> and <code style="color : white">**black**</code> / <code style="color : red">**red**</code> and <code style="color : yellow">**yellow**</code> (obviously this pattern is reversed in the other direction)
- all the time: 0 ohms (closed) between <code style="color : white">**black**</code> and <code style="color : yellow">**yellow**</code>

What I'm deducing from this is that <code style="color : yellow">**yellow**</code> is power, <code style="color : white">**black**</code> is ground, <code style="color : green">**green**</code> is the 3 switches with resistances of **0 ohms**, **6.8k ohms** and **22k ohms** each, and <code style="color : red">**red**</code> provides the rotary encoder signal with resistances of **∞ ohm** (open), **82k ohm**, **30k ohms** and **47k ohms** (with **30k** being the result of **82k** and **47k** in parallel). None of these can ever be activated simultaneously with the exception of pushing the inner shaft switch while turning it, which would be mutually exclusive since those 2 functions are on 2 separate wires, so that's the full extent of the options.

Cool!

Now how do I make that work with gpiozero? I was thinking of using an MCP3008 ADC, but I'm not sure how to convert values to button presses, and especially how to turn 4 resistance values into a working encoder functionality. I love the idea of being able to hook up 4 of these controls (12 switches and 4 rotary encoders!) to a Raspberry Pi by using just 4 pins of the GPIO, but I'm way out of my depth when it comes to gpiozero coding.

**Comment by lurch** [comment]
I don't have time to help you with the code-aspects of this, but if you're seeing "all the time: 0 ohms (closed) between black and yellow" then it sounds like black and yellow might **both** be ground? I.e. don't try connecting black to ground and yellow to power, as you'll short out your Pi! And it sounds like this is a passive component anyway, so it wouldn't seem to make sense that it needs "power"?

To measure a resistance using an ADC you'll need to use a potential-divider circuit, see e.g. https://devxplained.eu/en/blog/resistance-measurement and the "Reading Resistive Sensors" section of https://learn.sparkfun.com/tutorials/voltage-dividers/all

> but I'm way out of my depth when it comes to gpiozero coding.

In which case perhaps it's easier to simply use "separate" devices that are already natively supported by gpiozero? :shrug:

**Comment by GoremanX** [comment]
Well no, I wouldn't connect it straight from power to ground. There would have to be a resistor (10k ohm should be fine) on the ground line. I'm getting the impression you're not familiar with resistor ladder buttons.

There is no comparable control available on the market that would work directly with the Pi GPIO, and gpiozero does natively support the MCP3008 ADC that I was considering using, so I'm not sure what you're suggesting.

**Reply by lurch** [reply]
Sorry, I didn't mean to cause any offence. I was just suggesting that it may be easier to use a "standard" rotary-encoder https://gpiozero.readthedocs.io/en/latest/api_input.html#rotaryencoder than it is trying to modify gpiozero to work with your fancy resistor-ladder based device. But of course I have no idea what your end goals are, so maybe a normal rotary encoder indeed wouldn't be appropriate for your project.

**Reply by GoremanX** [reply]
I'm absolutely not offended in any way :) Just no further ahead in determining how to turn an MCU3008 reading into a button press. I only have 2 holes for controls to work with, and one of them is taken up by another control. This is a vintage 8-track player that I'm converting into a headless media player, and I can't modify the outward appearance. I need more external controls than just an encoder can provide, even a dual concentric encoder (which is rare as hens' teeth) isn't sufficient. The beauty of the momentary switches on the outer knobs is that they can be assigned different functions depending on whether I actuate them quickly or hold them. That's 4 potential functions just on the outer shaft.

I've done resistor ladder buttons on Arduino projects before and they worked flawlessly. That's how I was able to integrate 15 buttons and 2 knobs on this music player for my daughter, all with a single Arduino Feather board:
![IMG-20240204-195207987](https://github.com/gpiozero/gpiozero/assets/876310/be168c94-8853-478d-b1d0-3c372bbc3170)
There's 9 playlists accessible through the 9 buttons, and the LED of the currently playing playlist lights up the corresponding button. This was done with a different ADC chip that's no longer produced.

What I'm trying to accomplish here should be much simpler by comparison.

**Reply by lurch** [reply]
> What I'm trying to accomplish here should be much simpler by comparison.

Well, yes and no... :wink: 
Coding in Python for Gpiozero is probably easier than coding in C for Arduino; but in order for Gpiozero to look outwardly simple, it has a lot of internal complexity (just look at https://github.com/gpiozero/gpiozero/blob/master/gpiozero/mixins.py for example!)

I guess from a high-level point of view you want to use an ADC channel (in combination with your potential divider) to read the resistance of your "green" wire, and then use the scaled 0 to 1 value you get back from that to determine which of the 3 "buttons" was pressed, but I'm not quite sure how you'd hook that up to the rest of gpiozero's `Button` machinery. Possibly you'd need to create a custom python class, which just implements the same interface as the `Button` class? (or perhaps `ButtonBoard`?) There's currently no built-in support for resistor-ladder buttons in gpiozero.

I wonder if it might be easier to desolder the resistor-ladder PCB and just hook up the Pi's GPIO pins to the "8 terminals" that you mentioned in your initial comment? :shrug:

**Reply by GoremanX** [reply]
I considered desoldering it and using it as-is, I have enough GPIO pins left to accomplish that in this project. But I also REALLY like the idea of getting this implemented with a resistor ladder for future projects. There's some specific use cases (for example, vehicle steering wheel buttons and dials) where a 4-wire interface to support 12+ functions would be fantastic. It could be incorporated as part of the OEM clockspring, which often has a very limited number of conductors to work with. The only other way to accomplish this that I know of is a shift register, which I don't see supporting things like dials or encoders.
